source/common/upstream/health_discovery_service.cc

```diff
@@ -54,12 +94,69 @@ void HdsDelegate::onReceiveInitialMetadata(Http::HeaderMapPtr&& metadata) {
   UNREFERENCED_PARAMETER(metadata);
 }
 
+void HdsDelegate::processMessage(
+    std::unique_ptr<envoy::service::discovery::v2::HealthCheckSpecifier>&& message) {
+  ENVOY_LOG(debug, "New health check response message {} ", message->DebugString());
+  ASSERT(message);
+
+  for (auto& cluster_health_check : message->health_check()) {
+    // Create HdsCluster config
+    envoy::api::v2::core::BindConfig bind_config;
+
+    clusters_config_.emplace_back();
+    envoy::api::v2::Cluster& cluster_config = clusters_config_.back();
+
+    cluster_config.set_name(cluster_health_check.cluster_name());
+    cluster_config.mutable_connect_timeout()->set_seconds(ClusterTimeoutSeconds);
+    cluster_config.mutable_per_connection_buffer_limit_bytes()->set_value(
+        ClusterConnectionBufferLimitBytes);
+
+    // Add endpoints to cluster
+    for (auto& locality_endpoints : cluster_health_check.endpoints()) {
+      for (auto& endpoint : locality_endpoints.endpoints()) {
+        cluster_config.add_hosts()->MergeFrom(endpoint.address());
+      }
+    }
+
+    // TODO(lilika): Add support for optional per-endpoint health checks
+
+    // Add healthchecks to cluster
+    for (auto& health_check : cluster_health_check.health_checks()) {
+      cluster_config.add_health_checks()->MergeFrom(health_check);
+    }
+
+    ENVOY_LOG(debug, "New HdsCluster config {} ", cluster_config.DebugString());
+
+    // Create HdsCluster
+    hds_clusters_.emplace_back(new HdsCluster(runtime_, cluster_config, bind_config, store_stats,
+                                              ssl_context_manager_, secret_manager_, false,
+                                              info_factory_));
+
+    for (auto& health_check : cluster_config.health_checks()) {
+      health_checkers_.push_back(
```

Comment:
Arguably, `HdsCluster` could be response for owning and maintaining `health_checkers_`. Any thoughts here?